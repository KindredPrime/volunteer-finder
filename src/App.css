@import url('https://fonts.googleapis.com/css2?family=Karla&family=Sura&family=Varela+Round&display=swap');

* {
  box-sizing: border-box;
}

body {
  background-color: #1A4161;
  color: white;
  margin-bottom: 20px;
}

h1 {
  text-shadow: -3px 0 black, 0 3px black, 3px 0 black, 0 -3px black;
  font-size: 34px;
  line-height: 48px;
}

h2 {
  text-shadow: -2px 0 black, 0 2px black, 2px 0 black, 0 -2px black;
  font-size: 26px;
  line-height: 36px;
}

p, legend {
  text-shadow: -1px 0 black, 0 1px black, 1px 0 black, 0 -1px black;
}

h1, h2, legend {
  font-family: 'Varela Round', sans-serif;
}

legend {
  font-size: 18px;
  line-height: 24px;
}

p, label {
  font-family: 'Karla', sans-serif;
}

p, label {
  font-size: 16px;
  line-height: 24px;
}

p {
  text-align: left;
}

section {
  border-top: 3px solid #747A82;
  margin: 20px 0px 20px 0px;
}

section > * {
  margin-left: 5%;
  margin-right: 5%;
}

button {
  background-color: #FF9100;
  color: black;
  cursor: pointer;
}

button:disabled {
  background-color: #747A82;
}

a {
  text-decoration: none;
  color: white;
}

a, button {
  font-family: 'Sura', serif;
}

form {
  padding: 10px;
}

fieldset, legend {
  margin-left: auto;
  margin-right: auto;
}

label {
  margin-right: 10px;
}

form button {
  margin-top: 10px;
}

fieldset {
  border: none;
}

legend {
  font-weight: bold;
}

p.error {
  color: #FF9100;
}

ul {
  padding: 0;
}

li {
  list-style: none;
}

.App {
  margin-top: 75px;
  text-align: center;
}

.App.expanded {
  margin-top: 138px;
}

@media screen and (min-width: 750px) {
  .App.expanded {
    margin-top: 75px;
  }
}

@media screen and (min-width: 1200px) {
  .App, .App.expanded {
    margin-top: 0px;
    margin-left: 17%;
  }
}