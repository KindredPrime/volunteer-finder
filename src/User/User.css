.User__orgs, .User__events {
  list-style: none;
  padding: 0;
}

.User__orgs a, .User__events a {
  color: black;
}