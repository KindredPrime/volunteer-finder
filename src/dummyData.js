const dummyUsers = [
  {
    id: 1,
    email: 'email1@email.com',
    username: 'user1',
    password: 'password1',
    orgsAdded: [1],
    eventsAdded: [1, 2]
  },
  {
    id: 2,
    email: 'email2@email.com',
    username: 'user2',
    password: 'password2',
    orgsAdded: [2],
    eventsAdded: [3]
  },
  {
    id: 3,
    email: 'email3@email.com',
    username: 'user3',
    password: 'password3',
    orgsAdded: [3],
    eventsAdded: []
  },
  {
    id: 4,
    email: 'email4@email.com',
    username: 'user4',
    password: 'password4',
    orgsAdded: [],
    eventsAdded: []
  },
  {
    id: 5,
    email: 'email5@email.com',
    username: 'user5',
    password: 'password5',
    orgsAdded: [4],
    eventsAdded: [4]
  }
];

const dummyOrgs = [
  {
    id: 1,
    name: 'YMCA',
    website: 'https://www.ymca.com',
    phone: '1-800-555-7777',
    email: 'contact@ymca.com',
    address: '1 YMCA Street Real City, Real State',
    causes: [1, 2, 3],
    tags: [2, 3, 4, 6]
  },
  {
    id: 2,
    name: 'ASPCA',
    website: 'https://www.aspca.com',
    phone: '1-800-123-4567',
    email: 'contact@aspca.com',
    address: '1 ASPCA Street Real City, Real State',
    causes: [5, 2],
    tags: [2, 3, 4, 10]
  },
  {
    id: 3,
    name: 'National VOAD',
    website: 'https://www.nvoad.com',
    phone: '1-800-456-7890',
    email: 'contact@nvoad.com',
    address: '1 NVOAD Street Rivendell, Middle Earth',
    causes: [4, 2],
    tags: [2, 3, 4, 5, 7, 11]
  },
  {
    id: 4,
    name: 'National Air and Space Museum',
    website: 'https://www.airandspace.com',
    phone: '1-800-987-6543',
    email: 'contact@airandspace.com',
    address: '1 Space Street Air City, Outer Space',
    causes: [6, 7],
    tags: [1, 2, 3, 4, 6, 12]
  }
];

const dummyEvents = [
  {
    id: 1,
    name: 'Member Appreciation Day',
    organization: 1,
    location: 'Reston YMCA Center',
    date: 'February 20, 2020 00:00:00',
    duration: '4 hours',
    description: 'Greet YMCA members as they enter the facility and offer them snacks.  See https://www.website.com for more details.',
    causes: [2, 3],
    tags: [2, 4, 12]
  },
  {
    id: 2,
    name: 'Triathlon Assistance',
    organization: 1,
    location: 'Reston YMCA Center',
    date: 'January 27, 2020 00:00:00',
    duration: '4 hours',
    description: 'Help YMCA staff host their annual indoor triathlon: sign in triathlon participants, direct the participants to the different stages of the triathlon, and record times. See https://www.triathlon.com for more details.',
    causes: [2, 3],
    tags: [2, 3, 9]
  },
  {
    id: 3,
    name: 'Adoption Day',
    organization: 2,
    location: 'Times Square, New York City, New York',
    date: 'December 13, 2020 00:00:00',
    duration: '8 hours',
    description: 'Help the ASPCA run their annual adoption day for their local pet shelter.',
    causes: [5],
    tags: [2, 3, 9, 12]
  },
  {
    id: 4,
    name: 'Docent Training Class',
    organization: 4,
    location: '14390 Air and Space Museum Pkwy Chantilly, VA 20151',
    date: 'January 30, 2021 00:00:00',
    duration: '2 hours',
    description: 'This is a required class for anyone who wants to be a docent (tour guide) at the National Air and Space Museum.',
    causes: [6, 7],
    tags: [2, 3, 13]
  }
];

const dummyTags = [
  {
    id: 1,
    name: 'virtual'
  },
  {
    id: 2,
    name: 'in-person'
  },
  {
    id: 3,
    name: 'weekends'
  },
  {
    id: 4,
    name: 'weekdays'
  },
  {
    id: 5,
    name: 'nights'
  },
  {
    id: 6,
    name: 'group'
  },
  {
    id: 7,
    name: 'delivery'
  },
  {
    id: 8,
    name: 'cooking'
  },
  {
    id: 9,
    name: 'administration'
  },
  {
    id: 10,
    name: 'cleaning'
  },
  {
    id: 11,
    name: 'heavy lifting'
  },
  {
    id: 12,
    name: 'greeting'
  },
  {
    id: 13,
    name: 'training'
  }
];

const dummyCauses = [
  {
    id: 1,
    name: 'Youth'
  },
  {
    id: 2,
    name: 'Community'
  },
  {
    id: 3,
    name: 'Health'
  },
  {
    id: 4,
    name: 'Disaster Response'
  },
  {
    id: 5,
    name: 'Animals'
  },
  {
    id: 6,
    name: 'Museums'
  },
  {
    id: 7,
    name: 'Education'
  },
  {
    id: 8,
    name: 'Human Rights'
  },
  {
    id: 9,
    name: 'Arts'
  }
];

export {
  dummyUsers,
  dummyOrgs,
  dummyEvents,
  dummyTags,
  dummyCauses
};