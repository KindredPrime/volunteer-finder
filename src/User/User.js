import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import VolunteerContext from '../VolunteerContext';
import Nav from '../Nav/Nav';
import EventList from '../EventList/EventList';
import './User.css';

class User extends Component {
  static contextType = VolunteerContext;

  render() {
    const { user, orgs, events } = this.context;
    const { email, username, orgsAdded, eventsAdded } = user;

    const fullOrgsAdded = orgsAdded 
      ? orgsAdded.map((orgAdded) => orgs.find((org) => org.id === orgAdded)) 
      : [];

    const fullEventsAdded = eventsAdded 
      ? eventsAdded.map((eventAdded) => events.find((event) => event.id === eventAdded)) 
      : [];

    return (
      <div className="User">
        <Nav />

        <header>
          <h1>Account</h1>
        </header>

        <section>
          <header>
            <h2>Account Info</h2>
          </header>

          <p>Username: {username}</p>
          <p>Email: {email}</p>
        </section>

        <section>
          <header>
            <h2>Organizations Added</h2>
            <ul className="User__orgs">
              {fullOrgsAdded.map((org) => (
                <li key={org.id} className="User__org">
                  <Link to={`/org/${org.id}`}>{org.name}</Link>
                </li>
              ))}
            </ul>
          </header>
        </section>

        <section>
          <header>
            <h2>Events Added</h2>
          </header>

          <EventList events={fullEventsAdded} />
        </section>
      </div>
    );
  }
}

export default User;