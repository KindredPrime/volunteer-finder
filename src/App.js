import React, { Component } from 'react';
import { Route, Redirect } from 'react-router-dom';
import { dummyUsers, dummyOrgs, dummyCauses, dummyTags } from './dummyData';
import VolunteerContext from './VolunteerContext';
import Home from './Home/Home';
import Login from './Login/Login';
import SignUp from './SignUp/SignUp';
import Account from './Account/Account';
import Organization from './Organization/Organization';
import PublicUser from './PublicUser/PublicUser';
import OrgSearch from './OrgSearch/OrgSearch';
import AddOrg from './AddOrg/AddOrg';
import './App.css';

class App extends Component {
  static contextType = VolunteerContext;

  state = {
    user: null,
    users: dummyUsers,
    orgs: dummyOrgs,
    causes: dummyCauses,
    tags: dummyTags
  };

  loginUser = (username, password) => {
    const { users } = this.state;

    const user = users.find((elem) => (
      elem.username === username && elem.password === password
    ));

    if (user) {
      window.localStorage.setItem('userId', user.id);
      this.setState({
        user
      });
    }
    else {
      return 'Invalid username and password combination';
    }
  }

  logoutUser = () => {
    window.localStorage.removeItem('userId');
    this.setState({
      user: null
    });
  }

  signUpUser = (email, username, password) => {
    const { users } = this.state;

    const id = users.length + 1;
    users.push({
      id,
      email,
      username,
      password,
      orgsAdded: []
    });

    this.setState({
      users
    });

    this.loginUser(username, password);

    return id;
  }

  addOrg = (name, website, phone, email, address, causes, tags) => {
    const { user, users, orgs } = this.state;
    const id = orgs.length + 1;
    const org = {
      id,
      name,
      website: website || 'Not Available',
      phone: phone || 'Not Available',
      email: email || 'Not Available',
      address: address || 'Not Available',
      causes,
      tags
    };

    const newUser = {
      ...user,
      orgsAdded: [...user.orgsAdded, org.name]
    };
    const newUsers = users.map((elem) => {
      if (elem.id === user.id) {
        return newUser
      }

      return elem;
    });
    this.setState({
      user: newUser,
      users: newUsers,
      orgs: [...orgs, org]
    });

    return id;
  }

  componentDidMount() {
    const storedId = window.localStorage.getItem('userId');
    if (storedId) {
      const { users } = this.state;
      const user = users.find((elem) => elem.id === parseInt(storedId));
      this.setState({
        user
      });
    }
  }

  render() {
    const { user, users, orgs, causes, tags } = this.state;
    const contextValue = {
      user,
      users,
      orgs,
      causes,
      tags,
      loginUser: this.loginUser,
      logoutUser: this.logoutUser,
      signUpUser: this.signUpUser,
      addOrg: this.addOrg
    };

    return (
      <VolunteerContext.Provider value={contextValue}>
        <div className="App">
          <Route exact path="/" component={Home} />
          <Route path="/login" render={({ history }) => (
            !user ? <Login history={history} /> : <Redirect to="/account" />
          )} />
          <Route path="/signup" component={SignUp} />
          <Route path="/account" render={() => user ? <Account /> : <Redirect to="/login" />} />
          <Route path="/org/:id" component={Organization} />
          <Route path="/user/:id" component={PublicUser} />
          <Route path="/org-search" render={() => <OrgSearch pageLimit={10} />} />
          <Route path="/add-org" render={({ history }) => (
            window.localStorage.getItem('userId') ? <AddOrg history={history}/> : <Redirect to="/login" />
          )} />
        </div>
      </VolunteerContext.Provider>
    );
  }
}

export default App;
