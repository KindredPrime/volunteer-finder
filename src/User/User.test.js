import ReactDOM from 'react-dom';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import User from './User';
import App from '../App';
import { dummyUsers } from '../dummyData';

describe('User Component', () => {
  const origStoredId = window.localStorage.getItem('userId');

  beforeAll(() => {
    if (origStoredId) {
      window.localStorage.removeItem('userId');
    }
  });

  afterEach(() => {
    if (window.localStorage.getItem('userId')) {
      window.localStorage.removeItem('userId');
    }
  });

  afterAll(() => {
    if (origStoredId) {
      window.localStorage.setItem('userId', origStoredId);
    }
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <User />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    render(
      <MemoryRouter initialEntries={['/login']}>
        <App />
      </MemoryRouter>
    );

    userEvent.type(screen.getByLabelText('Username'), dummyUsers[0].username);
    userEvent.type(screen.getByLabelText('Password'), dummyUsers[0].password);
    userEvent.click(screen.getByRole('button', { name: 'Login' }));

    expect(document.body).toMatchSnapshot();
  });
});