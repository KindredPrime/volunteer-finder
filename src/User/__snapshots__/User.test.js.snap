// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Component renders the UI as expected 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="User"
      >
        <nav
          class="Nav"
        >
          <a
            href="/"
          >
            Home
          </a>
          <a
            aria-current="page"
            class="active"
            href="/user"
          >
            Account
          </a>
          <a
            href="/"
          >
            Logout
          </a>
        </nav>
        <header>
          <h1>
            Account
          </h1>
        </header>
        <section>
          <header>
            <h2>
              Account Info
            </h2>
          </header>
          <p>
            Username: 
            user1
          </p>
          <p>
            Email: 
            email1@email.com
          </p>
        </section>
        <section>
          <header>
            <h2>
              Organizations Added
            </h2>
            <ul
              class="User__orgs"
            >
              <li
                class="User__org"
              >
                <a
                  href="/org/1"
                >
                  YMCA
                </a>
              </li>
            </ul>
          </header>
        </section>
        <section>
          <header>
            <h2>
              Events Added
            </h2>
          </header>
          <div
            class="EventList"
          >
            <ul>
              <li>
                <a
                  class="EventList__event passed"
                  href="/event/1"
                >
                  Member Appreciation
                </a>
              </li>
              <li>
                <a
                  class="EventList__event passed"
                  href="/event/2"
                >
                  Triathlon Assistance
                </a>
              </li>
            </ul>
          </div>
        </section>
      </div>
    </div>
  </div>
</body>
`;
